### ISSUE TITLE
Answer's and question's feedback

### ISSUE THREAD

**Issue by PascalCrepey** [priism-center/shinyQuiz#7]
Shinyquiz cannot currently provide feedback for wrong answers. I guess it may be possible to do it through a user-defined grading function, but that seems a bit complex. 
A more natural way would be to add a new parameter to `add_choice()` and the other answer functions so that we may have the option to provide it (something like `add_choice(text, feedback = NULL, correct = FALSE)`) and display it accordingly. 

The GIFT format (https://docs.moodle.org/404/en/GIFT_format) also defines general feedback which could be provided to `create_question()`.

I am happy to contribute to this feature if you feel it's the way to go. 

FWIW, I have been recently working on [an R parser for GIFT](https://github.com/PascalCrepey/RGIFTParser), so I'm interested in extending this package to fit what the GIFT format allows to do.

**Comment by joemarlo** [comment]
This is a great idea. I agree that `add_choice` and equivalents are the right spot to integrate this in. How do you imagine the display of the feedback? Is a column on the results table sufficient?

Tangentially, do you think it is possible to integrate the GIFT parser into `shinyquiz` so a developer could generate a `quiz` object directly from the GIFT formatted string?

**Comment by PascalCrepey** [comment]
That would be the easiest way to implement it, but maybe not the most user-friendly. IMHO, it would be better to have the feedback appear when the green check or red cross appears on the submission. Encapsulating them in a well-positioned hidden <div> would work. That also means disabling the automatic switch to the next question and replacing it with a "next" button (so the user has enough time to read the feedback). However, this behavior must be optional as, in some circumstances, one may not want to give too much information.

Integrating the GIFT parser is a good idea. I was going to write some code to do it. When it's ready, I'll submit a PR. I also plan on submitting RGIFTParser to CRAN, but I must complete the grammar first (only MCQs have been implemented so far).

**Comment by joemarlo** [comment]
Let's start with using the results table as this should be straightforward to implement. Having the feedback on the same page will require some additional thought -- perhaps a modal could do the trick. I'll create a branch and draft a potential solution. Feel free to do the same and submit a PR if you'd like.

Yes, please submit a PR once the parser is ready and RGIFTParser is on CRAN. That could be a good addition.

**Comment by joemarlo** [comment]
@PascalCrepey I've pushed an initial draft at [feature/feedback branch](https://github.com/priism-center/shinyquiz/tree/feature/feedback). I believe this contains the right approach of integrating it into the class and state machine. It's not yet feature-completed; it only works with the `quizChoice` class (created via  `add_choice()`). Please give it a try (example script below). 

`add_choice()` has a natural place for the `feedback` arg since they directly map to a correct/wrong answer. `add_slider()`, `add_numeric()`, and `add_text()` are trickier since they could result in either a correct or wrong answer. Do you have any thoughts on the best way of implementing the argument in these?

```
devtools::load_all()

quiz <- create_quiz(
  create_question(
    'What is the capital of Illinois?',
    add_choice('Chicago', feedback = "Biggest city but not the capital"),
    add_choice('Paw Paw', feedback = 'Nope'),
    add_choice('Springfield', correct = TRUE, feedback = 'Great job!')
  ),
  create_question(
    'Which elements are gases at room temperature? Select all that apply.',
    add_choice('Hydrogen', correct = TRUE),
    add_choice('Mercury'),
    add_choice('Nitrogen', correct = TRUE),
    label = 'Select Hydrogen and Nitrogen'
  ),
  create_question(
    'At what temperature does water boil at sea level?',
    add_slider(min = 90, max = 150, default_position = 120, correct = 100),
    label = 'Select 100'
  ),
  options = set_quiz_options(question_heading = NULL)
)
preview_app(quiz)
```

**Comment by PascalCrepey** [comment]
FYI, the GIFT format provides two feedbacks for TRUE/FALSE questions ("when right" or "when wrong") ([see here](https://docs.moodle.org/404/en/GIFT_format#Feedback)). The way you implemented `add_slider(),` `add_numeric(),` and `add_text()` makes them look like an extension of a TRUE/FALSE question. So, I would say the best thing to do would be to have at least an additional input variable if you want shinyquiz to be fully compatible. The `feedback` variable could be considered by default the one used for wrong answers (feedback is mostly useful when you're wrong), and you may add a `feedback_right` input for `add_slider()`, `add_numeric()`, and `add_text()`.  
The other option would be to allow functions to be inputted: 

```r
add_slider(min = 90, max = 150, default_position = 120, correct = 100, 
  feedback = \(x, correct){
    if(x<correct) return("too cold")
    else if(x>correct) return("too warm")
    else return("Well done !")
  }
 )
``` 

I don't know which solution is the best in the long run and whether the balance "coding efforts" vs "usefulness" is good. But the latter option seems quite elegant... and similar to what you did for `grader`.  It will also be a nightmare for the GIFT -> shinyquiz translation but that's another issue...

**Comment by joemarlo** [comment]
Managing that balance of "coding efforts" and "usefulness" is important. Although the former is more about ease of maintainability. 

Your approach of providing a function to handle the feedback should work well and, as you mentioned, has precedence in `shinyquiz`. I'll dig into it this week and see if it can be implemented.

**Comment by joemarlo** [comment]
@PascalCrepey I've pushed an update to the [feature/feedback branch](https://github.com/priism-center/shinyquiz/tree/feature/feedback) which includes the function based approach. Please pull it down and test it out. There may still be a handful of edge cases that may cause it to crash so please let me know if you run into anything.

Here is an example script:
```
devtools::load_all()

quiz <- create_quiz(
  create_question(
    'What is the capital of Illinois?',
    add_choice('Chicago', feedback = "Biggest city but not the capital"),
    add_choice('Paw Paw', feedback = 'Nope'),
    add_choice('Springfield', correct = TRUE, feedback = 'Great job!')
  ),
  create_question(
    'At what temperature does water boil at sea level?',
    add_numeric(
      correct = 100,
      feedback = function(x, correct){
        if (x < correct) return("Too low!")
        if (x > correct) return("Too High!")
        return("Just right")
      }
    ),
    label = 'Select 100'
  ),
  create_question(
    'At what temperature does water boil at sea level?',
    add_slider(
      min = 90, 
      max = 150, 
      default_position = 120, 
      correct = 100,
      feedback = function(x, correct){
        if (x < correct) return("Too low!")
        if (x > correct) return("Too High!")
        return("Just right")
      }
      ),
    label = 'Select 100'
  ),
  create_question(
    'At what temperature does water boil at sea level?',
    add_text(
      correct = "100",
      feedback = function(x, correct) {
        if (setequal(x, correct)) return("Great!")
        return("Try again!")
      }
    ),
    label = 'Select 100'
  ),
  options = set_quiz_options(question_heading = NULL)
)
preview_app(quiz)
```
